File: solutions/rust/0020.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
        let map: HashMap<char, char> = HashMap::from([
            ('(', ')'),
            ('{', '}'),
            ('[', ']')
        ]);
        for c in s.chars() {
            if map.contains_key(&c) {
                stack.push(c);
            } else if stack.len() > 0 && map[&stack[stack.len() - 1]] == c {
                stack.pop();
            } else {
                return false;
            }
        }
        stack.is_empty()
    }
}

File: solutions/go/0020.go
func isValid(s string) bool {
    stack := []rune{}
    for _, char := range s {
        if char == '(' || char == '{' || char == '[' {
            stack = append(stack, char)
        } else if char == ')' || char == '}' || char == ']' {
            if len(stack) == 0 {
                return false
            }
            top := stack[len(stack)-1]
            if (char == ')' && top != '(') || (char == '}' && top != '{') || (char == ']' && top != '[') {
                return false
            }
            stack = stack[:len(stack)-1]
        }
    }
    return len(stack) == 0
}